pkg/commands/chgm/chgm.go

```diff
@@ -0,0 +1,137 @@
+//go:generate mockgen -source $GOFILE -destination ./mock/interfaces.go -package mock
+type ChgmService interface {
+	// AWS
+	ListStopInstancesEvents() ([]*cloudtrail.Event, error)
+	ListRunningInstances(infraId string) ([]*ec2.Instance, error)
+	ListStoppedInstances(infraId string) ([]*ec2.Instance, error)
+	// OCM
+	GetClusterDeployment(clusterID string) (*hivev1.ClusterDeployment, error)
+	GetClusterInfo(identifier string) (*v1.Cluster, error)
+}
+
+// Chgm is an implementation of the Interface, and adds functionality above it
+// to craete Chgm you can use a mock, or fill it with IsRunningImpl
+type Chgm struct {
+	ChgmService
+}
+
+// isUserAllowedToStop will verify if a user is allowed to stop instances.
+// as this is the thing that might change the most it is at the top.
+// additionally it is private as I don't see anyone using this outside of AreAllInstancesRunning
+func isUserAllowedToStop(username, infraID string) bool {
+	//TODO: what is the best ordering for this? (from the most common to the most rare)
+
+	// pulled by:
+	// 1. logging (via osdctl account cli) into an aws cluster
+	// 2. running the command "aws iam list-users --query 'Users[?starts_with(UserName,`<INFRA_ID>`)].UserName'"
+	// 3. trimming the infra id from the front and the uuid from the back
+	operatorIamNames := []string{
+		"cloud-credential-operator-iam-ro",
+		"cloud-ingress-operator-credentials-aw",
+		"instance",
+		"managed-velero-operator-iam-credentia",
+		"openshift-image-registry",
+		"openshift-ingress",
+		"openshift-machine-api-aws",
+		"sre-ebs-iops-reporter-aws-credentials",
+		"sre-stuck-ebs-vols-aws-credentials",
+		"velero-iam-credentials",
+	}
+	for _, operatorIamName := range operatorIamNames {
+		// I added an additional '-' in the end to make sure it's the same as close as I can to the aws username (cannot guess the uuid)
+		operatorIamName = fmt.Sprintf("%s-%s-", infraID, operatorIamName)
+		if strings.HasPrefix(username, operatorIamName) {
+			return true
+		}
+	}
+
+	switch {
+	// all of these following cases didn't have any users connected to them
+	//case strings.HasPrefix(username, "RH-SRE"):
+	//case strings.HasPrefix(username, "osdAdmin-"):
+	//case strings.HasPrefix(username, "ManagedOpenShift-Support-"):
+	// this one on the other hand did :)
+	case strings.HasPrefix(username, "osdManagedAdmin-"):
+		return true
+	default:
+		return false
+	}
+}
+
+// AreAllInstancesRunning will check all of the instances under the externalID are running and return false if any of them are not
+func (i Chgm) AreAllInstancesRunning(externalID string) (bool, error) {
+	cluster, err := i.GetClusterInfo(externalID)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve cluster info for %s: %w", externalID, err)
+	}
+	id := cluster.ID()
+
+	cd, err := i.GetClusterDeployment(id)
+	if err != nil {
+		return false, fmt.Errorf("could not retrieve Cluster Deployment for %s: %w", id, err)
+	}
+	infraID := cd.Spec.ClusterMetadata.InfraID
+
```

Comment:
just as a reminder once you got the internal ClusterID you have to do the assume role chain at that point. As we are having our base account credentials, and need to get into the customer account.
something like:
`
supportRoleArn := ocm.GetSupportRoleARN(id)
client, err = i.awsClient.AssumeRole(jumprole)    // fetched from config
ccsClient, err =  client.AssumeRole(supportRoleArn)
`

Reply:
is that something I need to take care of, or should be provided by the person creating the AwsClient?

Reply:
its ok for now, we can refactor it at one point. I think it is just required for solving that ticket. But yes I would put it somewhere else. But It has to come after the internalID was fetched which is now the beginning of your function. As this becomes very long, we can maybe put it immediately after we build up the client for the base/main account. somewhere after this line: https://github.com/openshift/configuration-anomaly-detection/blob/932ab869d1114f9b4867794a53fc2b4fb31c651c/cadctl/cmd/cluster-missing/cluster-missing.go#L41